### PR TITLE
GSYE-178: Enable grid fees accounting for DoF requirements

### PR DESCRIPTION
**Pull request by Hassan754**
We used to add grid fees to the orders' prices when they get forwarded to another market.
Now that the orders have other `price` source of truth (the ones in the requirements), when we add fees to the order.price we also have to add fees to all `price` key in all `order.requirements` list (otherwise offer seller may end up getting less than it requested)

### CODE REVIEW

File: src/gsy_e/models/strategy/market_agents/two_sided_engine.py
Comment (hannesdiedrich) [inline src/gsy_e/models/strategy/market_agents/two_sided_engine.py:55]: Can you please explain why the `original_price` should be part of an requirement?

Comment (hannesdiedrich) [reply src/gsy_e/models/strategy/market_agents/two_sided_engine.py:55]: I am sorry, but I think I also do not get the concept of a `price` requirement. Can you please explain when this could be useful?
When I post a bid like `Bid(energy=10, price=2)`, I already set the price, don't I? Why would I set an price requirement?

Comment (Hassan754) [reply src/gsy_e/models/strategy/market_agents/two_sided_engine.py:55]: Sure, let me elaborate.
The goal of the degrees of freedom is to allow buyers to have flexibility in the buying process, for instance I can post a bid `Bid(energy=10, price=2, requirements=[{"preferred_partners": ["Hannes"], "price": 1}])`

Comment (Hassan754) [reply src/gsy_e/models/strategy/market_agents/two_sided_engine.py:55]: So the bid in this case will be sold to "Hannes" for a lower price.
And as for the `original_price` requirement, it serves the same goal as the original_price attribute of the orders (denoting the price before adding/substracting grid fees)

Comment (hannesdiedrich) [reply src/gsy_e/models/strategy/market_agents/two_sided_engine.py:55]: I see, thank you! So if the order was not forwarded yet, the "original_price" would not be part of the requirement, and from the moment of first forwarding, this member is in the requirement, right?
Do we have a validator for the requirements (I had a quick look but could not find it)? I mean to prohibit to let the user set the member "original_price" in order to make the grid fee calculation work. What do you think?

File: src/gsy_e/models/strategy/market_agents/two_sided_engine.py
Comment (hannesdiedrich) [inline src/gsy_e/models/strategy/market_agents/two_sided_engine.py:61]: NIT: I found the bracket placing weird. As it works as you proposed, you can leave it if you like. 
```suggestion
                    updated_requirement["original_price"] / energy) * energy)
```

File: src/gsy_e/models/strategy/market_agents/two_sided_engine.py
Comment (Hassan754) [reply src/gsy_e/models/strategy/market_agents/two_sided_engine.py:55]: > if the order was not forwarded yet, the "original_price" would not be part of the requirement, and from the moment of first forwarding, this member is in the requirement

That's correct!
Regarding the validator, I can now see its importance, thanks for foreseeing it!

**Comment by Hassan754** [comment]
> I see that you only adapted the grid_fee calculation to the bids. Is there any reason not to do it for the offers?

This is because the price requirement is only supported for bids, see [here](https://github.com/gridsingularity/gsy-framework/blob/4d6009981bae87ad63aae81f4a9bfba1d72eaf9f/gsy_framework/matching_algorithms/requirements_validators.py#L107)

File: src/gsy_e/models/strategy/market_agents/two_sided_engine.py
Comment (Hassan754) [reply src/gsy_e/models/strategy/market_agents/two_sided_engine.py:55]: @hannesdiedrich the validator needs some thorough thinking and designing to cover all corner cases, if you agree I'll add this in follow up PR after merging both this and [the other](https://github.com/gridsingularity/gsy-e/pull/1375) PRs

Comment (hannesdiedrich) [reply src/gsy_e/models/strategy/market_agents/two_sided_engine.py:55]: Ok, I am also fine with adding a ticket for this and postpone it a bit.

**Comment by hannesdiedrich** [comment]
> > I see that you only adapted the grid_fee calculation to the bids. Is there any reason not to do it for the offers?
> 
> This is because the price requirement is only supported for bids, see [here](https://github.com/gridsingularity/gsy-framework/blob/4d6009981bae87ad63aae81f4a9bfba1d72eaf9f/gsy_framework/matching_algorithms/requirements_validators.py#L107)

Your link is not pointing correctly AFAIU. But as you say, if this is only supported for bids, all good.

**Review by hannesdiedrich — APPROVED** [review]
LGTM

File: src/gsy_e/models/strategy/market_agents/two_sided_engine.py
Comment (hannesdiedrich) [reply src/gsy_e/models/strategy/market_agents/two_sided_engine.py:55]: @Hassan754 Please add a note into the ticket to tell us to consider some time for the `thorough thinking and designing to cover all corner cases` Thanks!

File: src/gsy_e/models/market/two_sided.py
Comment (fievelk) [inline src/gsy_e/models/market/two_sided.py:93]: Nit: given what the `update_incoming_bid_with_fee` method is doing (just selecting one of the two rates) these lines seem a bit redundant. They're basically doing:
```python
updated_price = updated_requirement["price"] or updated_requirement["original_price"]
```
This could be ok if the implementation of `update_incoming_bid_with_fee()` might change in the future. Regardless, these lines seem a bit redundant somehow and, since you wrote the `BaseClassGridFees` interface, you might be the right person to see if this can be changed or if it's better to keep it as it is.

Comment (fievelk) [reply src/gsy_e/models/market/two_sided.py:93]: One more thing: `GridFees.update_incoming_bid_with_fee()` is not doing what the name implies. It is not updating anything, it's just selecting which rate to apply.

Comment (fievelk) [reply src/gsy_e/models/market/two_sided.py:93]: First thing popping in my head: given the name, I would probably expect `update_incoming_bid_with_fee()` to take a `Bid` as an argument and modify it. This would internally do what you're doing in `_update_requirements_prices()`. Just an idea, you might find other better ones.

Comment (Hassan754) [reply src/gsy_e/models/market/two_sided.py:93]: @fievelk I agree with your proposals and indeed these modules need to be further refactored, actually I did not write them I only fixed some stylings warnings in this [PR](https://github.com/gridsingularity/gsy-e/pull/1373).
After an offline discussion with Spyros, we agreed not to alter the logic to avoid introducing bugs, especially because we'll revisit and reimplement the entire fees modules after getting rid of the hierarchical setups we currently support (One more point to discuss in the upcoming workshop I assume)

File: src/gsy_e/models/market/two_sided.py
Comment (fievelk) [inline src/gsy_e/models/market/two_sided.py:84]: Nit: I think we could change the [init method](https://github.com/gridsingularity/gsy-framework/blob/4d6009981bae87ad63aae81f4a9bfba1d72eaf9f/gsy_framework/data_classes.py#L58) to always provide an empty list for `Bid.requirements`, so that we can avoid this check.

File: src/gsy_e/models/strategy/market_agents/two_sided_engine.py
Comment (fievelk) [reply src/gsy_e/models/strategy/market_agents/two_sided_engine.py:55]: > I see, thank you! So if the order was not forwarded yet, the "original_price" would not be part of the requirement, and from the moment of first forwarding, this member is in the requirement, right?

I think it might also be worth adding the `original_price` at the very beginning, before the order is forwarded. This might simplify our life and we could avoid needing to check every time if `"original_price"` is present or not.

**Review by fievelk — APPROVED** [review]
I left some comments. It might be good to address the complexity of `update_incoming_bid_with_fee()`, but I'll leave an LGTM and leave it up to you.

**Pull request merged**
